### PR TITLE
{WIP} [Ripple] Fixing Ripple to allow correct theming of selected+highlighted states in Chips.

### PR DESCRIPTION
**Pull request by galiak11**
# Description
Fixing Ripple to allow correct theming of selected and highlighted states in Chips.

1. Adding "allowsSelection" to MDCChipView, which enables toggling selection of the RippleView.
2. Moving rippleStateForControlState: method from MDCChipView to MDCStatefulRippleView,  so it can be shared by multiple components.
3. Removing state management from MDCChipCollectionCell, directing all states setters/getters to its MDCChipView instance.
4. To align with UIKit/UIControl behavior, we remove the default colors for Dragged and Selected states in MDCStatfulRippleView. The Default is now taken from the Normal state.

Additional Fixes to ChipCells, StatefulRippleView and Chips:

5. StatefulRippleView: Fix case where when allowsSelection is turned on in the ripple, but the collectionView is using the notion of allowMultipleSelection = NO, we don't want that if the Ripple is selected and is then tapped on again, that it will have 2 selected layers. This makes sure that the additional ripple layer dissolves right after if the ripple is already selected.

6. ChipCells: When Ripple is enabled I have removed the notion of the ChipCell maintaining its own state system and passing that to its subview the chipView. This causes complications like when the cell is selected, UIKit turns all of its subviews (including the chipView) to highlighted. Therefore we don't call super for setSelected and setHighlighted, but rather just passthrough the state to the chipView which is the core component.

# Issues
b/133759923 - Incorrect ripple color for the highlighted + selected state in Chips
PRs:  #7609   |   #7613   |   #7616

### CODE REVIEW

File: components/Chips/src/MDCChipView.m [hunk 1]
```diff
@@ -276,6 +276,14 @@ - (void)setEnableRippleBehavior:(BOOL)enableRippleBehavior {
   }
 }
 
+- (BOOL)allowsSelection {
+  return !self.enableRippleBehavior || self.rippleView.allowsSelection;
```
Comment (yarneo) [inline components/Chips/src/MDCChipView.m]: would this mean that if ripple isn't enabled then allowSelection will always return YES, is this intended?

**Review by yarneo — COMMENTED** [review]
Overall the changes look fine. 
I would like to see though these changes being cut up to smaller PRs, which will make the changes more succinct, easier to review, and in cases where a revert may be needed, we can revert smaller chunks rather than all of this work.
Also, I think there still needs to be some work in MDCChipView and possibly the themers, to remove setting the selected background color if ripple behavior is enabled, and instead set the ripple color for selected and selected+highlighted.

File: components/Chips/src/MDCChipView.m (hunk 1 above)
Comment (galiak11) [reply components/Chips/src/MDCChipView.m]: Yes. My assumption is that this is the desired/current behavior of ink in chips, but I 'm actually not sure. Is it not the current behavior in ink?

**Review by galiak11 — COMMENTED** [review]
I've tested all these changes together, so it is hard to know what affect it will have if I took parts of it out.   I also ran a global presubmit to verify this change.  

I could move the rippleStateForControlState method back to the ChipsView without too much impact. Can you be more specific if there are other parts that you would like taken out?

The changes you mentioned to themers/theming extensions depend on the desired theming outcome - and will be done in a separate pr/cl as needed.

Note that there is a second part to this fix in PR #7586 - which adds a disabled state to Ripple and fixes theming of disabled+selected chips.

File: components/Chips/src/MDCChipView.m (hunk 1 above)
Comment (yarneo) [reply components/Chips/src/MDCChipView.m]: I believe this is only the case when using an MDCChipCollectionCell, whereas an MDCChipView goes from highlighted then back to normal. You will have to check and confirm though on the types of chips and in the Dragons examples. If there are chips that dont go into selected after tapping, then this should facilitate that here too.

**Review by yarneo — CHANGES_REQUESTED** [review]
I would like to see the MDCStatefulRippleView changes be separated from the ChipView and ChipCollectionCell changes. I would also like to see within the Chip changes, the exposing of allowSelection in a separate PR as well. This will end up with 3 smaller PRs.

Once split, each PR should have just one item/task that it is achieving, which should be summarized in the PRs. This will help clean up your description and make it easier for the reviewer to know exactly the changes you are proposing and why. As an example:
PR Title: "[Chips] Addition of allowSelection API to enable/disable selection" Description: "Similarly to UICollectionView's API, by enabling allowSelection on the Chip, once it is tapped, the chip state will go from normal to highlighted to selected, rather than from normal to highlighted then back to normal. This is wanted behavior in certain types of chips, and by exposing the MDCStatefulRippleView's allowSelection, we are providing this behavior toggle in the API"

Lastly, it would be great to have unit tests covering the additional code as I see both the ripple change and the chip changes risky.

If this is time sensitive, I am happy to land a patch release later this week.

File: components/Chips/src/MDCChipView.m (hunk 1 above)
Comment (galiak11) [reply components/Chips/src/MDCChipView.m]: Because this is a Ripple API, when enableRippleBehavior is not enabled, the behavior is unpredictable.

**Comment by jverkoey** [comment]
Closing this due to inactivity, but please reopen if we intend to pursue it again.

**Pull request closed without merging**